### PR TITLE
IGNITE-21921 Sql. Cover SQL E071-01 (Basic query expressions, UNION DISTINCT table operator) feature by tests

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# name: union_all_projection_pushdown.test
+# description: Test projection pushdown in UNION ALL
+# group: [setops]
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test:3]: please add a reference to the E071-01 SQL feature

File: modules/sql-engine/src/integrationTest/sql/union/basic_union.test [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+query II rowsort
+SELECT 1, 2 UNION ALL SELECT i, i FROM a
+----
+NULL	NULL
+100	100
+1	2
+
+
+
+
```
Comment (AMashenkov) [inline modules/sql-engine/src/integrationTest/sql/union/basic_union.test:80]:
```suggestion
```

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test
Comment (korlov42) [inline modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test:1]: all test names in the directory starts with `test_` prefix, let's follow this pattern. Besides, there is already `test_subquery_union.test`, so let's rename old script and new script to `test_union_of_subquery.test` and `test_union_in_subquery.test` accordingly to better reflect tests' purposes.

**Review by korlov42 — COMMENTED** [review]
I see that every tests uses table with integer attributes. Let's add test cases fro types combinations

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java [hunk 3]
```diff
@@ -39,6 +39,8 @@ public RehashingPartitionExtractor(
     /** {@inheritDoc} */
     @Override
     public int partition(RowT row) {
+        assert fields.length != 0 : "Unexpected partition calculation call";
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java:42]: as i can see for now it possible to call constructor with empty fields, i append this assertion for signalling that this method never called with empty fields.
Seems we obtain this situation here :
`    Int2ObjectMap<NodeWithConsistencyToken> assignments(List<String> nodeNames) {
        if (!(this instanceof PartitionedTarget)) {
            return Int2ObjectMaps.emptyMap();
        }`
when this instance of OneOfTarget

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test (hunk 1 above)
Comment (zstan) [reply modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test]: done

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test
Comment (zstan) [reply modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test:1]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java:42]: we should not create partition extractor with empty fields

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test
Comment (korlov42) [reply modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test:1]: what about second part of my suggestion?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/RehashingPartitionExtractor.java:42]: I've fixed this assertion in https://github.com/apache/ignite-3/pull/3579/files#diff-a16f16888749cc53d94ad940a53a9d44ea982dd5455f17b89cc0b0772cffd7a9, it worth to merge latest main into your branch

File: modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test
Comment (zstan) [reply modules/sql-engine/src/integrationTest/sql/subquery/table/union_all_projection_pushdown.test:1]: done

**Review by korlov42 — APPROVED** [review]

**Pull request merged**